### PR TITLE
logical_count: improve memory usage

### PR DESCRIPTION
**Pull request by HashidaTKS**
Overall, I have changed `logical_count` to immediately `unref` shards or `close` temporary objects that are no longer referred like `logical_range_filter`.

TODO:

- [x] Refactor
  * Reconsider name of method, class and so on
  * Separate common method with `logical_range_filter` into a common class
    * e.g. `detect_time_classify_types`
    * If they have not match in common, I will skip this.
- [x] More test
    * Especially performance test

### CODE REVIEW

File: test/command/suite/sharding/logical_count/log/select.test [hunk 1]
```diff
@@ -5,6 +5,11 @@ plugin_register sharding
 table_create Logs_20150709 TABLE_NO_KEY
 column_create Logs_20150709 timestamp COLUMN_SCALAR Time
 
+load --table Logs_20150709
+[
+{"timestamp": "2015-02-05 13:49:00"}
+]
```
Comment (HashidaTKS) [inline test/command/suite/sharding/logical_count/log/select.test:11]: I have added this because the expected log `#|d| [logical_count][select] <Logs_20150709>: no range index` is not written if `Logs_20150709` is empty.

Currently, an empty table is skipped before logging that log.
https://github.com/groonga/groonga/pull/1341/files#diff-d3fb57a047ca92c9fc19d900aa1a906a31bc3519eb7c26927516b229705c27a4R290

File: test/command/suite/sharding/logical_count/log/range_index.test [hunk 2]
```diff
@@ -6,6 +6,10 @@ table_create Logs_20150709 TABLE_NO_KEY
 column_create Logs_20150709 timestamp COLUMN_SCALAR Time
 table_create Times_20150709 TABLE_PAT_KEY Time
 column_create Times_20150709 timestamp_index COLUMN_INDEX Logs_20150709 timestamp
+load --table Logs_20150709
+[
+{"timestamp": "2015-02-05 13:49:00"}
+]
```
Comment (HashidaTKS) [inline test/command/suite/sharding/logical_count/log/range_index.test:13]: I have added this because the expected log `#|d| [logical_count][range-index] <Logs_20150709>: range index is available` is not written if `Logs_20150709` is empty.

Currently, an empty table is skipped before logging that log.
https://github.com/groonga/groonga/pull/1341/files#diff-d3fb57a047ca92c9fc19d900aa1a906a31bc3519eb7c26927516b229705c27a4R290

File: test/command/suite/sharding/logical_count/log/cover_type_all.test [hunk 3]
```diff
@@ -6,6 +6,10 @@ table_create Logs_20150709 TABLE_NO_KEY
 column_create Logs_20150709 timestamp COLUMN_SCALAR Time
 table_create Times_20150709 TABLE_PAT_KEY Time
 column_create Times_20150709 timestamp_index COLUMN_INDEX Logs_20150709 timestamp
+load --table Logs_20150709
+[
+{"timestamp": "2015-02-05 13:49:00"}
+]
```
Comment (HashidaTKS) [inline test/command/suite/sharding/logical_count/log/cover_type_all.test:12]: I have added this because the expected log `#|d| [logical_count][select] <Logs_20150709>: covered` is not written if `Logs_20150709` is empty.

Currently, an empty table is skipped before logging that log.
https://github.com/groonga/groonga/pull/1341/files#diff-d3fb57a047ca92c9fc19d900aa1a906a31bc3519eb7c26927516b229705c27a4R290

File: plugins/sharding/logical_count.rb [hunk 4]
```diff
@@ -1,277 +1,527 @@
...
-            counter.count_pre(shard, shard_range)
-          end
-          total += counter.count
-          unless have_shard
-            message =
-              "[logical_count] no shard exists: " +
-              "logical_table: <#{enumerator.logical_table}>: " +
-              "shard_key: <#{enumerator.shard_key_name}>"
-            raise InvalidArgument, message
+        register("logical_count",
```
Comment (kou) [inline plugins/sharding/logical_count.rb:4]: Could you fix indent for easy to review?

File: test/command/suite/sharding/logical_count/log/cover_type_all.expected [hunk 5]
```diff
@@ -8,10 +8,15 @@ table_create Times_20150709 TABLE_PAT_KEY Time
 [[0,0.0,0.0],true]
 column_create Times_20150709 timestamp_index COLUMN_INDEX Logs_20150709 timestamp
 [[0,0.0,0.0],true]
+load --table Logs_20150709
+[
+{"timestamp": "2015-02-05 13:49:00"}
```
Comment (kou) [inline test/command/suite/sharding/logical_count/log/cover_type_all.expected:13]: This table should have only `2015-07-09` timestamps.

File: plugins/sharding/logical_count.rb (hunk 4 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: Sorry, I have (auto) formatted `logical_count.rb`.

File: test/command/suite/sharding/logical_count/log/cover_type_all.expected (hunk 5 above)
Comment (HashidaTKS) [reply test/command/suite/sharding/logical_count/log/cover_type_all.expected]: Right, fix to use 2015-07-09 timestamps.

File: plugins/sharding/logical_count.rb [hunk 6]
```diff
@@ -2,42 +2,27 @@ module Groonga
...
-          total += counter.count
-          unless have_shard
-            message =
-              "[logical_count] no shard exists: " +
-              "logical_table: <#{enumerator.logical_table}>: " +
-              "shard_key: <#{enumerator.shard_key_name}>"
-            raise InvalidArgument, message
-          end
+          counter = Counter.new(context)
+          count_result = counter.count
```
Comment (kou) [inline plugins/sharding/logical_count.rb:20]: Why did you rename this from `total`?
It seems that `count_result` isn't better than `total`.

File: plugins/sharding/logical_count.rb [hunk 7]
```diff
@@ -57,79 +42,296 @@ def cache_key(input)
       end
 
       class ShardCountContext
```
Comment (kou) [inline plugins/sharding/logical_count.rb:44]: It seems that new `ShardCountContext` is similar to `GroongaSharding::LogicalRangeFilterCommand::ExecuteContext`.
Can we unify common implementation?

File: plugins/sharding/logical_count.rb [hunk 8]
```diff
@@ -57,79 +42,296 @@ def cache_key(input)
...
       end
 
       class Counter
+        def initialize(context)
+          @context = context
+        end
+
+        def count
+          have_shard = false
+          have_result_set = false
```
Comment (kou) [inline plugins/sharding/logical_count.rb:144]: Not used.

File: plugins/sharding/logical_count.rb [hunk 9]
```diff
@@ -57,79 +42,296 @@ def cache_key(input)
...
+              "[logical_count] no shard exists: " +
+              "logical_table: <#{enumerator.logical_table}>: " +
+              "shard_key: <#{enumerator.shard_key_name}>"
+            raise InvalidArgument, message
+          end
+          total
+        end
+
+        private
+        def each_shard_executor(&block)
```
Comment (kou) [inline plugins/sharding/logical_count.rb:162]: This is similar to `Groonga::Sharding::LogicalRangeFilterCommand::Executer#each_shard_executor`.
Can we unify common implementation?

File: test/command/suite/sharding/logical_count/log/select.expected [hunk 10]
```diff
@@ -4,10 +4,14 @@ table_create Logs_20150709 TABLE_NO_KEY
...
+load --table Logs_20150709
+[
+{"timestamp": "2015-07-09 20:00:00"}
+]
+[[0,0.0,0.0],1]
 log_level --level debug
 [[0,0.0,0.0],true]
 logical_count Logs timestamp   --min "2015/07/09 00:00:00"   --max "2015/07/09 12:00:00"
 [[0,0.0,0.0],0]
-#|d| [logical_count][select] <Logs_20150709>: no range index
```
Comment (HashidaTKS) [inline test/command/suite/sharding/logical_count/log/select.expected:11]: Currently the debug log `no range index` is not outputted.
I will fix them in another PR because we need to modify StreamShardExecutor (https://github.com/groonga/groonga/blob/master/plugins/sharding/stream_shard_executor.rb#L74) for fixing it and it affects to not only `logical_count` but also `logical_range_filter`.

File: plugins/sharding/logical_count.rb [hunk 11]
```diff
@@ -56,179 +41,134 @@ def cache_key(input)
...
+          if @filter or @post_filter
+            decide_use_range_index(false, "need filter",
+                                __LINE__, __method__)
+          elsif @cover_type == :all
+            decide_use_range_index(false, "covered",
+                                __LINE__, __method__)
+          else
+            decide_use_range_index(true, "range index is available",
+                                __LINE__, __method__)
+          end
```
Comment (HashidaTKS) [inline plugins/sharding/logical_count.rb:132]: Currently the debug log `no range index` is not outputted.
I will fix it in another PR because we need to modify StreamShardExecutor (https://github.com/groonga/groonga/blob/master/plugins/sharding/stream_shard_executor.rb#L74) for fixing it and it affects to not only `logical_count` but also `logical_range_filter`.

File: plugins/sharding/logical_count.rb (hunk 6 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: I have fixed it.

File: plugins/sharding/logical_count.rb (hunk 8 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: I have removed it.

File: plugins/sharding/logical_count.rb (hunk 9 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: I have unified common implementation with `logical_range_filter` in #1342, #1343, #1345.

File: plugins/sharding/logical_count.rb (hunk 7 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: I have unified common implementation with logical_range_filter in https://github.com/groonga/groonga/pull/1342, https://github.com/groonga/groonga/pull/1343, https://github.com/groonga/groonga/pull/1345.

**Comment by HashidaTKS** [comment]
I am testing performance but ready for review of implementation.
I will write test results to this PR.

File: plugins/sharding/logical_count.rb [hunk 12]
```diff
@@ -56,179 +41,134 @@ def cache_key(input)
...
-        attr_reader :cover_type
-        attr_reader :range_index
-        attr_accessor :table
-        def initialize(shard, cover_type, range_index)
-          @shard = shard
-          @cover_type = cover_type
-          @range_index = range_index
-          @table = shard.table
+      class ShardCountContext < StreamExecuteContext
+        attr_reader :order
```
Comment (kou) [inline plugins/sharding/logical_count.rb:45]: It seems that we should be defined in `StreamExecuteContext`.

How about passing `order` to `StreamExecuteContext#initialize`?

File: plugins/sharding/logical_count.rb [hunk 13]
```diff
@@ -56,179 +41,134 @@ def cache_key(input)
...
         end
 
-        private
-        def log_use_range_index(use, table_name, reason, line, method)
-          message = "[logical_count]"
-          if use
-            message << "[range-index]"
-          else
-            message << "[select]"
+        def execute
```
Comment (kou) [inline plugins/sharding/logical_count.rb:84]: It seems that most of this code similar to `LogicalRangeFilterCommand::ShardExecutor#execute`.
Can we unify it and this?

File: plugins/sharding/logical_count.rb [hunk 14]
```diff
@@ -56,179 +41,134 @@ def cache_key(input)
...
+          ensure_filtered
+
+          if @range_index
+            return count_n_records_in_range
           end
```
Comment (kou) [inline plugins/sharding/logical_count.rb:89]: Why is this done before `return 0 if @filtered_result_sets.empty?`?
Can we return `0` here when we have no result sets?

File: plugins/sharding/logical_count.rb [hunk 15]
```diff
@@ -56,179 +41,134 @@ def cache_key(input)
...
+            decide_use_range_index(true, "range index is available",
+                                __LINE__, __method__)
+          end
+        end
 
-          if context.cover_type == :all and @filter.nil?
-            if @post_filter and @dynamic_columns.have_filtered?
-              filtered_table = context.table.select_all
+        def execute_filter(range_index)
+          return if range_index
```
Comment (kou) [inline plugins/sharding/logical_count.rb:136]: Umm... "`execute_filter` does nothing if there is a range index" will confuse developers...
Can we use another approach?

File: plugins/sharding/logical_count.rb (hunk 13 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb:84]: Okay.
I have create a new PR #1346 to unity it and this.

File: plugins/sharding/logical_count.rb (hunk 15 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: I have fixed it. How about the current approach?

- Removed this `return if range_index`.
- Modify to check `@range_index` before filtering in `ShardCountContext.execute`.

File: plugins/sharding/logical_count.rb (hunk 14 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb:88]: This is related to https://github.com/groonga/groonga/pull/1341#discussion_r839166934.
We don't have any result set when @range_index is defined.

File: plugins/sharding/logical_count.rb (hunk 12 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb:45]: Please let me work on this in another PR because it also affects to `logical_range_filter`.

File: plugins/sharding/logical_count.rb (hunk 14 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb:88]: >  We don't have any result set when @range_index is defined.
In `logical_count`, `@range_index` is defined only when we don't need to execute `filter` and simply get `index_cursor.count
` (https://github.com/groonga/groonga/pull/1341/files#diff-d3fb57a047ca92c9fc19d900aa1a906a31bc3519eb7c26927516b229705c27a4R191)

File: plugins/sharding/logical_count.rb (hunk 12 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb:45]: #1347

**Comment by HashidaTKS** [comment]
Performance test results.

Condition:
* 10M record per shard.
* Search 30 shards. 
* Filter with indexed columns

Sorry for my lack of details but I cannot write detailed logs or a schema and so on because it is containing private information...

**case: `GRN_ENABLE_REFERENCE_COUNT=yes`**

Execution time
|Condition|Execution time|
|--|--|
|After changed|47 sec|
|Before changed|54 sec|

Memory usage
image

**case: `GRN_ENABLE_REFERENCE_COUNT=no`**
Execution time
|Condition|Execution time|
|--|--|
|After changed|42 sec|
|Before changed|42 sec|

Memory usage

image

File: plugins/sharding/logical_count.rb [hunk 16]
```diff
@@ -56,179 +41,115 @@ def cache_key(input)
...
-          if use
-            message << "[range-index]"
-          else
-            message << "[select]"
+          total = 0
+          each_result_set do |result_set|
+            unless @post_filter.nil?
+              result_set = apply_post_filter(result_set)
+              @temporary_tables << result_set
+            end
```
Comment (kou) [inline plugins/sharding/logical_count.rb:98]: Can we move this to `StreamShardExecutor`?

File: plugins/sharding/logical_count.rb (hunk 13 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb:82]: #1346 is merged

File: plugins/sharding/logical_count.rb (hunk 11 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb:110]: This is resolved by #1349

File: plugins/sharding/logical_count.rb (hunk 7 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: They are merged.

File: plugins/sharding/logical_count.rb (hunk 12 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: #1347 is merged

File: plugins/sharding/logical_count.rb (hunk 11 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb:110]: This is resolved by #1349

File: test/command/suite/sharding/logical_count/log/select.expected (hunk 10 above)
Comment (HashidaTKS) [reply test/command/suite/sharding/logical_count/log/select.expected]: This is resolved by #1349

File: plugins/sharding/logical_count.rb (hunk 16 above)
Comment (HashidaTKS) [reply plugins/sharding/logical_count.rb]: #1348 is merged
